### DISCUSSION TITLE
Using Bash output inside a script

### DISCUSSION THREAD

**Discussion by theianjones** [johnlindquist/kit#922] [Q&A]
I am looking to port this shell script into scriptkit: https://github.com/willplatnick/calendar-tana

I want to call the [ical-buddy-json.sh](https://github.com/WillPlatnick/calendar-tana/blob/main/ical-buddy-json.sh) script as is. This script looks sufficiently complex to leave it written as is. 

How can I access the output of this script (in some variable maybe?) and parse the result in javascript?

**Comment by johnlindquist** [comment]
## Get Shell Script Output Once Complete

The most common operation is to wait for the script to finish, then just grab the `stdout` "Standard Output" ([Standard Streams](https://en.wikipedia.org/wiki/Standard_streams)) as a string. `zx` makes this rather trivial:

> Note: There is also a `stderr` "Standard Error" property which may have more info about what's happening in the process. While it's usually used to detect errors, it is often used for general info like installation progress, etc. So "error" is kind of a misnomer.

> Note: There are no real rules around `stdout` and `stderr`, they're both just streams of text which can also include special characters for colors, clearing the terminal, etc. It's impossible to write code that handles every script. You'll have to handle them on a script by script basis.

```js
// Name: Process Shell Script Output

import "@johnlindquist/kit"

// --- Create a shell script to run
// ~/.kenv/tmp/process-shell-script-output/example.sh
let shellScriptPath = tmpPath("example.sh")
await writeFile(
  shellScriptPath,
  `
echo "Hello there"
cd ~/.kit
ls
`
)

// Note: We're using $ (AKA "zx") because it's popular, but `exec` works much the same

// Make shell script executable
await $`chmod +x ${shellScriptPath}`

// ---

// Get _ALL_ of the output
let { stdout } = await $`${shellScriptPath}`

dev(stdout) // inspect output
```

[Open process-shell-script-output in Script Kit](https://scriptkit.com/api/new?name=process-shell-script-output&url=https://gist.githubusercontent.com/johnlindquist/abb369febf5e7686a1937f7157178496/raw/3d68af2758a5a4c17625fa0209851f144cf01884/process-shell-script-output.ts")

## Get Shell Script Output "Line by Line"

Many script starts long-running processes where you want to see the output in real-time rather than waiting for it to complete.

For this, we'll use the [Node Streams](https://nodejs.org/api/stream.html) APIs:

> Note: A script spawned with zx/exec is considered a "child process". The main script is the parent process. They have separate `stdout` streams which can be merged together and pass messages between each other. The streaming APIs handle passing messages between the main and child processes.

```js
// Name: Process Shell Output as Stream

import "@johnlindquist/kit"

// --- Create a shell script to run -----------------
// `tmpPath` will store the file here:
// ~/.kenv/tmp/process-shell-script-output/example.sh
let shellScriptPath = tmpPath("example.sh")

// We're sleeping between each command to mock a long-running process
let scriptContents = `
echo "Hello there"
sleep 1
cd ~/.kit
ls | grep .js$
sleep 1
ls | grep .md$
sleep 1
ls | grep .json$
`
await writeFile(shellScriptPath, scriptContents)

// Note: We're using $ (AKA "zx") because it's popular. `exec` works much the same
// Make shell script executable
await $`chmod +x ${shellScriptPath}`
// --------------------------------------------------

let output = ``
let divPromise = div()

// Get a Stream of Output
let { stdout } = $`${shellScriptPath}`

// Create a stream where we can pipe the output.
// You can think of `write` as a handler
// You _must_ call `next()` to move things forward
let writableStream = new Writable({
  write: async (chunk, encoding, next) => {
    let line = chunk.toString()
    output += line + "\n\n"
    setDiv(md(output))

    next()
  },
})

stdout.pipe(writableStream)

await divPromise
```

[Open process-shell-output-as-stream in Script Kit](https://scriptkit.com/api/new?name=process-shell-output-as-stream&url=https://gist.githubusercontent.com/johnlindquist/bcc1f2e5c492ec0e1d209b146a710901/raw/94fcbfa1ee14b03959d11ff8e90ecbba1e1056b9/process-shell-output-as-stream.ts")

**Comment by Ambushfall** [comment]
# Windows Compatible .ps1 .bat .sh version
## SH is WSL Only for now

### win32-run-script.js
```ts
// Name: Run .bat/.ps1/.sh
// Description: Process Output to Kit via stream

import "@johnlindquist/kit"

// @ts-expect-error
import { backToMainShortcut, highlightJavaScript } from "@johnlindquist/kit"

// --- Create a shell script to run -----------------
// `tmpPath` will store the file here:
// ~/.kenv/tmp/process-shell-script-output/example.*

// Note: linux shell will only work with WSL or you can provide the Args for ps1 using the .sh extension if you have gitbash

const fileName = "example"

const selectedLang = {
  name: '',
  args: '',
  ext: '',
  echo: '',

  set setVal(keyValueList: string[]) {
    this[keyValueList[0]] = keyValueList[1];
  }
};

const objGen = (_lang: string, _ext: string, _args?: string) => {
  _args = _args ? _args : ""
  return {
    name: _lang,
    description: `Run Script using ${_lang}`,
    value: _lang,
    id: _ext,
    arguments: _args,
    preview: async () => highlightJavaScript(tmpPath(`${fileName}.${_ext}`))
  }
}

const LangOptions = [
  objGen("PowerShell",
    "ps1",
    "powershell -NoProfile -NonInteractive –ExecutionPolicy Bypass -File "
  ), objGen("Batch", "bat"),
  objGen("Bash", "sh"
  )]

const promptEditor = ["yes", "no"]

const selectedValue = await arg("Use editor?", promptEditor
)

const useEditor = selectedValue === "yes" ? true : false

// define select options

await arg({
  placeholder: "Select Scripting Language...",
  enter: "Select",
  shortcuts: [backToMainShortcut],
  onChoiceFocus: async (input, { focused }) => {
    selectedLang.setVal = ["args", focused["arguments"]]
    selectedLang.setVal = ["ext", focused.id]
    selectedLang.setVal = ["name", focused.name]
    selectedLang.setVal = ["echo", selectedLang.ext == "bat" ? "@echo off" : ""]
  }
}, LangOptions)



let shellScriptPath = tmpPath(`${fileName}.${selectedLang.ext}`)

const editorConfig = {
  hint: `Write code for ${selectedLang.ext} file.`,
  description: 'Save to Run',
  onInputSubmit: async (input: any) => {
    selectedLang.ext == "sh" ? submit(`${input}
  exit`) : submit(input)
  }
}

// Using ping to simulate waiting for a long process and because it's natively supported across PS and Bat files
// Note: If you use a code that would natively not run in bat like "ls" it will 
let scriptContents = useEditor ? await editor(editorConfig) : `${selectedLang.echo}
echo "hello"
echo "Done"
${selectedLang.ext == "sh" ? "exit" : ""}
`

await writeFile(shellScriptPath, scriptContents);

// Just a wrapper to highlight with code in PS style
const codeWrapper = (string: string, extension: any) => `
\`\`\`${extension}
${string}
\`\`\`
`

let output = ``

// This is used to avoid kit window closing on process exit
let divPromise = div()

const outHandler = async (out: any) => {
  output += `${out}\n`;
  setDiv(
    await highlight(`${codeWrapper(output, selectedLang.ext)}`)
  )
};

// Note: We have to use this janky way of executing PS as it would launch in Notepad or fail entirely.
const execArgs = selectedLang.ext == "sh" ? `cd ${tmpPath()} && bash ${fileName}.sh` : `${selectedLang.args}${shellScriptPath}`

// inspect(execArgs)
let { stdout } = execLog(execArgs, outHandler)

setAlwaysOnTop(true);
setIgnoreBlur(true);

await divPromise
```

**Reply by johnlindquist** [reply]
@Ambushfall the `execLog` helper does something similar where the second argument is a function/callback you want to log to

**Reply by johnlindquist** [reply]
Oh yeah, I tossed this together about longer running scripts as well:

https://www.youtube.com/watch?v=mzOZGVUd8Ws

**Reply by Ambushfall** [reply]
@johnlindquist 

Thanks for the answer!

I've made use of that one and have updated the answer.

Not really sure what I think about the log functionality being locked away, alas it is what it is.

**Reply by johnlindquist** [reply]
@Ambushfall I tried to make the pro/sponsor features (debugger and logger) geared towards people who would use Script Kit _a lot_, but don't help with the code/community.

I'm working on a post about "contribute/help with Script Kit to get a free pro account". I'll definitely get you hooked up tomorrow when I'm back and my laptop

**Reply by Ambushfall** [reply]
@johnlindquist  Do you guys have or plan a discord server any time in the future?
I'd love to be a part of that!

**Reply by johnlindquist** [reply]
@Ambushfall Oh, I thought I sent you an invite in another thread:

https://discord.gg/3nxaTHM2
